tokens-accounting/src/cache.rs

```diff
@@ -0,0 +1,229 @@
+
+    fn get_token_data(&self, id: &TokenId) -> Result<Option<TokenData>, Self::Error> {
+        match self.data.token_data.get_data(id) {
+            accounting::GetDataResult::Present(d) => Ok(Some(d.clone())),
+            accounting::GetDataResult::Deleted => Ok(None),
+            accounting::GetDataResult::Missing => {
+                Ok(self.parent.get_token_data(id).map_err(|_| Error::ViewFail)?)
+            }
+        }
+    }
+
+    fn get_circulating_supply(&self, id: &TokenId) -> Result<Option<Amount>, Self::Error> {
+        let parent_supply = self.parent.get_circulating_supply(id).map_err(|_| Error::ViewFail)?;
+        let local_delta = self.data.circulating_supply.data().get(id).cloned();
+        combine_amount_delta(&parent_supply, &local_delta).map_err(Error::AccountingError)
+    }
+}
+
+impl<P> FlushableTokensAccountingView for TokensAccountingCache<P> {
+    type Error = Error;
+
+    fn batch_write_tokens_data(
+        &mut self,
+        delta: TokensAccountingDeltaData,
+    ) -> Result<crate::data::TokensAccountingDeltaUndoData, Self::Error> {
+        self.data.merge_with_delta(delta)
+    }
+}
+
+impl<P: TokensAccountingView> TokensAccountingOperations for TokensAccountingCache<P> {
+    fn issue_token(&mut self, id: TokenId, data: TokenData) -> Result<TokenAccountingUndo, Error> {
+        if self.get_token_data(&id)?.is_some() {
+            return Err(Error::TokenAlreadyExist(id));
+        }
+
+        if self.get_circulating_supply(&id)?.is_some() {
+            return Err(Error::TokenAlreadyExist(id));
+        }
+
+        let undo_data = self
+            .data
+            .token_data
+            .merge_delta_data_element(id, accounting::DataDelta::new(None, Some(data)))?;
+
+        Ok(TokenAccountingUndo::IssueToken(IssueTokenUndo {
+            id,
+            undo_data,
+        }))
+    }
+
+    fn mint_tokens(
+        &mut self,
+        id: TokenId,
+        amount_to_add: Amount,
+    ) -> Result<TokenAccountingUndo, Error> {
+        let token_data = self.get_token_data(&id)?.ok_or(Error::TokenDataNotFound(id))?;
+        let circulating_supply = self.get_circulating_supply(&id)?.unwrap_or(Amount::ZERO);
+
+        match token_data {
+            TokenData::FungibleToken(data) => match data.total_supply() {
+                TokenTotalSupply::Fixed(limit) => {
+                    let expected_circulating_supply =
+                        (circulating_supply + amount_to_add).ok_or(Error::AmountOverflow)?;
+                    if expected_circulating_supply > *limit {
+                        return Err(Error::MintExceedsSupplyLimit(amount_to_add, *limit, id));
+                    }
+                }
+                TokenTotalSupply::Lockable => {
+                    if data.is_locked() {
+                        return Err(Error::CannotMintFromLockedSupply(id));
+                    }
+                }
+                TokenTotalSupply::Unlimited => { /* do nothing */ }
+            },
```

Comment:
Shouldn't we enforce the i128::MAX limit here? Or do you just expect it to be enforced below in the add_unsigned function?

Reply:
yes I see it as an implementation detail of the accounting